From 03313f7e7c530294c9f6b51a5f49ca87875cf786 Mon Sep 17 00:00:00 2001
From: Bui Trung Kien <kien.bui@exacttechnology.com>
Date: Tue, 25 Jun 2024 13:55:27 +0700
Subject: [PATCH] HW-1655: Add method to verify the write/read data for mcuboot
 dfu target - Applied callback to flash sync - Re-write data after detecting
 inconsistency data

Signed-off-by: Bui Trung Kien <kien.bui@exacttechnology.com>
---
 .../dfu/dfu_target/src/dfu_target_mcuboot.c   | 140 +++++++++++++++++-
 subsys/dfu/dfu_target/src/dfu_target_stream.c |   2 +-
 .../net/lib/fota_download/src/fota_download.c |   9 ++
 3 files changed, 142 insertions(+), 9 deletions(-)

diff --git a/subsys/dfu/dfu_target/src/dfu_target_mcuboot.c b/subsys/dfu/dfu_target/src/dfu_target_mcuboot.c
index 34da33fb1..c4baec7b4 100644
--- a/subsys/dfu/dfu_target/src/dfu_target_mcuboot.c
+++ b/subsys/dfu/dfu_target/src/dfu_target_mcuboot.c
@@ -90,6 +90,11 @@ static size_t stream_buf_len;
 static size_t stream_buf_bytes;
 static uint8_t curr_sec_img;
 
+static uint8_t *stream_verify_buf;
+static size_t stream_verify_buf_len;
+static size_t stream_verify_write_len;
+static size_t stream_verify_offset;
+
 bool dfu_target_mcuboot_identify(const void *const buf)
 {
 	/* MCUBoot headers starts with 4 byte magic word */
@@ -112,6 +117,120 @@ int dfu_target_mcuboot_set_buf(uint8_t *buf, size_t len)
 	return 0;
 }
 
+int dfu_target_mcuboot_set_verify_buf(uint8_t *buf, size_t len)
+{
+	if (buf == NULL) {
+		return -EINVAL;
+	}
+
+	if (!IS_ALIGNED_32(buf)) {
+		return -EINVAL;
+	}
+
+	stream_verify_buf = buf;
+	stream_verify_buf_len = len;
+
+	return 0;
+}
+
+static int dfu_target_mcuboot_recover_data(uint8_t *buf, size_t len, size_t offset)
+{
+	int rc = 0;
+	struct flash_pages_info info;
+	struct stream_flash_ctx *ctx = dfu_target_stream_get_stream();
+	rc = flash_get_page_info_by_offs(ctx->fdev, offset, &info);
+	LOG_DBG("Page info %d", info.size);
+	off_t offset_sector = offset - (offset % info.size);
+	uint32_t remain_data_size = offset - offset_sector;
+	if (remain_data_size == 0) {
+		rc = flash_erase(ctx->fdev, offset_sector, info.size);
+		if (rc) {
+			LOG_ERR("[Recover] Can't erase the offset data");
+			return -EINVAL;
+		}
+	} else {
+		uint8_t *tmp_buf = (uint8_t *)k_malloc(remain_data_size);
+		if (tmp_buf == NULL) {
+			LOG_ERR("[Recover] Can't allocate memory for recover data");
+			return -ENOMEM;
+		}
+		rc = flash_read(ctx->fdev, offset_sector, tmp_buf, remain_data_size);
+		if (rc) {
+			LOG_ERR("[Recover] Can't read the offset data");
+			return -EINVAL;
+		}
+		rc = flash_erase(ctx->fdev, offset_sector, info.size);
+		if (rc) {
+			LOG_ERR("[Recover] Can't erase the offset data");
+			return -EINVAL;
+		}
+		rc = flash_write(ctx->fdev, offset_sector, tmp_buf, remain_data_size);
+		if (rc) {
+			LOG_ERR("[Recover] Can't write the recover data back");
+			return -EINVAL;
+		}
+		k_free(tmp_buf);
+	}
+	rc = flash_write(ctx->fdev, offset, buf, len);
+	if (rc != 0) {
+		LOG_ERR("[Recover] Error in writing flash err %d", rc);
+		return rc;
+	}
+	return 0;
+}
+
+static int dfu_target_mcuboot_verify(off_t write_addr, uint8_t *read_buf, size_t read_len,
+				     const uint8_t *verify_buf, size_t verify_len)
+{
+
+	struct stream_flash_ctx *ctx = dfu_target_stream_get_stream();
+	for (int i = 0; i < read_len; i++) {
+		read_buf[i] = ~read_buf[i];
+	}
+
+	int rc = flash_read(ctx->fdev, write_addr, read_buf, read_len);
+	if (rc != 0) {
+		LOG_ERR("flash read failed: %d", rc);
+		return rc;
+	}
+	LOG_HEXDUMP_DBG(read_buf, 32, "READ");
+	LOG_HEXDUMP_DBG(verify_buf, 32, "VERIFY");
+	if (memcmp(read_buf, verify_buf, read_len) != 0) {
+		LOG_ERR("[Recover] Still invalid data");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int dfu_target_mcuboot_callback_sync(uint8_t *buf, size_t len, size_t offset)
+{
+	LOG_DBG("Write 0x%08x - %d", offset, len);
+	if (stream_verify_offset < stream_verify_write_len) {
+		if (memcmp(buf, &stream_verify_buf[stream_verify_offset], len) != 0) {
+			LOG_WRN("[Recover] Invalid data - Need to recover");
+			uint8_t *recover_buf = &stream_verify_buf[stream_verify_offset];
+			size_t recover_len = len;
+			LOG_HEXDUMP_DBG(buf, 32, "READ");
+			LOG_HEXDUMP_DBG(recover_buf, 32, "VERIFY");
+			int rc = dfu_target_mcuboot_recover_data(recover_buf, recover_len, offset);
+			if (rc) {
+				LOG_ERR("[Recover] Can't recover current status %d", rc);
+				return rc;
+			} else {
+				rc = dfu_target_mcuboot_verify(offset, buf, len, recover_buf,
+							       recover_len);
+				if (rc) {
+					LOG_ERR("Verfied failed");
+					return rc;
+				}
+				LOG_WRN("[Recover] Recovered data successfully");
+			}
+		}
+		stream_verify_offset += len;
+	}
+	return 0;
+}
+
 int dfu_target_mcuboot_init(size_t file_size, int img_num, dfu_target_callback_t cb)
 {
 	ARG_UNUSED(cb);
@@ -138,14 +257,14 @@ int dfu_target_mcuboot_init(size_t file_size, int img_num, dfu_target_callback_t
 		return -EFAULT;
 	}
 
-	err = dfu_target_stream_init(&(struct dfu_target_stream_init){
-		.id = target_id_name[img_num],
-		.fdev = flash_dev,
-		.buf = stream_buf,
-		.len = stream_buf_len,
-		.offset = secondary_address[img_num],
-		.size = secondary_size[img_num],
-		.cb = NULL });
+	err = dfu_target_stream_init(
+		&(struct dfu_target_stream_init){.id = target_id_name[img_num],
+						 .fdev = flash_dev,
+						 .buf = stream_buf,
+						 .len = stream_buf_len,
+						 .offset = secondary_address[img_num],
+						 .size = secondary_size[img_num],
+						 .cb = dfu_target_mcuboot_callback_sync});
 	if (err < 0) {
 		LOG_ERR("dfu_target_stream_init failed %d", err);
 		return err;
@@ -170,6 +289,11 @@ int dfu_target_mcuboot_offset_get(size_t *out)
 int dfu_target_mcuboot_write(const void *const buf, size_t len)
 {
 	stream_buf_bytes = (stream_buf_bytes + len) % stream_buf_len;
+	__ASSERT_NO_MSG(len <= stream_verify_buf_len);
+
+	stream_verify_write_len = len;
+	stream_verify_offset = 0;
+	memcpy(stream_verify_buf, buf, len);
 
 	return dfu_target_stream_write(buf, len);
 }
diff --git a/subsys/dfu/dfu_target/src/dfu_target_stream.c b/subsys/dfu/dfu_target/src/dfu_target_stream.c
index ed905e828..027023e27 100644
--- a/subsys/dfu/dfu_target/src/dfu_target_stream.c
+++ b/subsys/dfu/dfu_target/src/dfu_target_stream.c
@@ -112,7 +112,7 @@ int dfu_target_stream_init(const struct dfu_target_stream_init *init)
 	current_id = init->id;
 
 	err = stream_flash_init(&stream, init->fdev, init->buf, init->len,
-				init->offset, init->size, NULL);
+				init->offset, init->size, init->cb);
 	if (err) {
 		LOG_ERR("stream_flash_init failed (err %d)", err);
 		return err;
diff --git a/subsys/net/lib/fota_download/src/fota_download.c b/subsys/net/lib/fota_download/src/fota_download.c
index 0c290313e..36b336dd3 100644
--- a/subsys/net/lib/fota_download/src/fota_download.c
+++ b/subsys/net/lib/fota_download/src/fota_download.c
@@ -37,6 +37,7 @@ static struct k_work_delayable  dlc_with_offset_work;
 static int socket_retries_left;
 #ifdef CONFIG_DFU_TARGET_MCUBOOT
 static uint8_t mcuboot_buf[CONFIG_FOTA_DOWNLOAD_MCUBOOT_FLASH_BUF_SZ] __aligned(4);
+static uint8_t mcuboot_verify_buf[CONFIG_DOWNLOAD_CLIENT_HTTP_FRAG_SIZE] __aligned(4);
 #endif
 static enum dfu_target_image_type img_type;
 static enum dfu_target_image_type img_type_expected = DFU_TARGET_IMAGE_TYPE_ANY;
@@ -52,6 +53,8 @@ static atomic_t flags;
 static enum fota_download_error_cause error_state = FOTA_DOWNLOAD_ERROR_CAUSE_NO_ERROR;
 static bool initialized;
 
+extern int dfu_target_mcuboot_set_verify_buf(uint8_t *buf, size_t len);
+
 static void send_evt(enum fota_download_evt_id id)
 {
 	__ASSERT(id != FOTA_DOWNLOAD_EVT_PROGRESS, "use send_progress");
@@ -607,6 +610,12 @@ int fota_download_init(fota_download_callback_t client_callback)
 		LOG_ERR("%s failed to set MCUboot flash buffer %d", __func__, err);
 		return err;
 	}
+
+	err = dfu_target_mcuboot_set_verify_buf(mcuboot_verify_buf, sizeof(mcuboot_verify_buf));
+	if (err) {
+		LOG_ERR("%s failed to set MCUboot verify flash buffer %d", __func__, err);
+		return err;
+	}
 #endif
 
 	use_smp_dfu_target = false;
-- 
2.42.0.windows.1

